### PR TITLE
Bulk User Import API

### PR DESCRIPTION
**Pull request by hiranya911**
Introduces the following new APIs for importing users to Firebase Auth in bulk (1000 users per API call):

```
FirebaseAuth.importUsers(List<UserImportRecord> users)
FirebaseAuth.importUsers(List<UserImportRecord> users, UserImportOptions options)
FirebaseAuth.importUsersAsync(List<UserImportRecord> users)
FirebaseAuth.importUsersAsync(List<UserImportRecord> users, UserImportOptions options)
```

Also add the `UserImportHash` and `UserImportResult` APIs.

go/firebase-admin-user-import

### CODE REVIEW

File: src/main/java/com/google/firebase/auth/hash/Scrypt.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+import com.google.firebase.auth.UserImportHash;
+import java.util.Map;
+
+/**
+ * Represents the Scrypt password hashing algorithm. This is the
+ * <a href="https://github.com/firebase/scrypt">modified Scrypt algorithm</a> used by
+ * Firebase Auth. See {@link StandardScrypt} for the standard Scrypt algorithm. Can be used as an
+ * instance of {@link com.google.firebase.auth.UserImportHash} when importing users.
+ */
+public final class Scrypt extends UserImportHash {
```
Comment (schmidt-sebastian) [inline src/main/java/com/google/firebase/auth/hash/Scrypt.java]: Can we make this inherit BasicHash and re-use `setRounds()`? On that note, should `BasicHash` be renamed to better reflect what it does?

File: src/main/java/com/google/firebase/auth/internal/UploadAccountResponse.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.auth.internal;
+
+import com.google.api.client.util.Key;
+import java.util.List;
+
+public class UploadAccountResponse {
```
Comment (schmidt-sebastian) [inline src/main/java/com/google/firebase/auth/internal/UploadAccountResponse.java]: Please add a some JavaDoc snippet, even for internal classes.

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 3]
```diff
@@ -433,6 +438,66 @@ public void testCustomTokenWithClaims() throws Exception {
...
+            .setMemoryCost(14)
+            .build())).get();
+    assertEquals(1, result.getSuccessCount());
+    assertEquals(0, result.getFailureCount());
+
+    try {
+      UserRecord savedUser = auth.getUserAsync(randomId).get();
+      assertEquals(userEmail, savedUser.getEmail());
+      String idToken = signInWithPassword(userEmail, "password");
+      assertTrue(!Strings.isNullOrEmpty(idToken));
```
Comment (schmidt-sebastian) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: assertFalse

File: src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java [hunk 4]
```diff
@@ -252,6 +252,158 @@ public void testDeleteUser() throws Exception {
...
+        .setUid("user2")
+        .setPasswordHash("password".getBytes())
+        .build();
+
+    List<ImportUserRecord> users = ImmutableList.of(user1, user2);
+    try {
+      FirebaseAuth.getInstance().importUsersAsync(users);
+      fail("No error thrown for missing hash option");
+    } catch (IllegalArgumentException expected) {
+      // expected
```
Comment (schmidt-sebastian) [inline src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java]: Please verify the error message. That will also make it super obvious what this test is verifying.

File: src/main/java/com/google/firebase/auth/ImportUserRecord.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    /**
+     * Adds a user provider to be associated with this user.
+     *
+     * @param provider A non-null {@link UserProvider}.
+     * @return This builder.
+     */
+    public Builder addUserProvider(@NonNull UserProvider provider) {
+      this.userProviders.add(provider);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/ImportUserRecord.java]: Done

File: src/main/java/com/google/firebase/auth/FirebaseUserManager.java [hunk 6]
```diff
@@ -257,4 +269,37 @@ private void handleHttpError(HttpResponseException e) throws FirebaseAuthExcepti
...
+        if (user.hasPassword()) {
+          hasPassword = true;
+        }
+        usersBuilder.add(user.getProperties(jsonFactory));
+      }
+      this.users = usersBuilder.build();
+
+      if (hasPassword) {
+        checkArgument(options != null && options.getHash() != null,
+            "UserImportHash option is required when at least one user has a password");
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/FirebaseUserManager.java]: Done

File: src/main/java/com/google/firebase/auth/UserImportOptions.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      return this;
+    }
+
+    /**
+     * Builds a new {@link UserImportOptions}.
+     *
+     * @return A non-null {@link UserImportOptions}.
+     */
+    public UserImportOptions build() {
+      return new UserImportOptions(this);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/UserImportOptions.java]: Done

File: src/main/java/com/google/firebase/auth/UserImportOptions.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+   *
+   * @return A {@link UserImportOptions.Builder} instance.
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  Map<String, Object> getProperties() {
+    ImmutableMap.Builder<String, Object> properties = ImmutableMap.builder();
+    if (hash != null) {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/UserImportOptions.java]: Done

File: src/main/java/com/google/firebase/auth/UserImportResult.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      }
+    }
+    this.users = users;
+    this.errors = errorsBuilder.build();
+  }
+
+  /**
+   * Returns the number of users that were imported successfully.
+   *
+   * @return an integer (possibly zero).
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/UserImportResult.java]: Done

File: src/main/java/com/google/firebase/auth/UserImportResult.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  /**
+   * A list of {@link ErrorInfo} instances describing the errors that were encountered during
+   * the import. Length of this list is equal to the return value of {@link #getFailureCount()}.
+   *
+   * @return A non-null list (possibly empty).
+   */
+  @NonNull public List<ErrorInfo> getErrors() {
+    return errors;
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/UserImportResult.java]: This is already an instance of `ImmutableList`. Changed the private field declaration to make it clear.

File: src/main/java/com/google/firebase/auth/UserRecord.java [hunk 11]
```diff
@@ -232,29 +229,44 @@ public UpdateRequest updateRequest() {
...
     checkArgument(password.length() >= 6, "password must be at least 6 characters long");
   }
 
-  private static void checkCustomClaims(Map<String,Object> customClaims) {
+  static void checkCustomClaims(Map<String,Object> customClaims) {
     if (customClaims == null) {
       return;
     }
     for (String key : customClaims.keySet()) {
+      checkArgument(!"".equals(key), "Claim names must not be empty");
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/UserRecord.java]: Done

File: src/main/java/com/google/firebase/auth/UserRecord.java [hunk 12]
```diff
@@ -232,29 +229,44 @@ public UpdateRequest updateRequest() {
...
+    checkArgument(phoneNumber.startsWith("+"),
         "phone number must be a valid, E.164 compliant identifier starting with a '+' sign");
   }
 
+  static void checkPhotoUrl(String photoUrl) {
+    checkArgument(!Strings.isNullOrEmpty(photoUrl), "photoUrl cannot be null or empty");
+    try {
+      new URL(photoUrl);
+    } catch (MalformedURLException e) {
+      throw new IllegalArgumentException("malformed photoUrl string", e);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/UserRecord.java]: It does include it. For example if the input string is `not a url`, we get:

```
java.lang.IllegalArgumentException: malformed photoUrl string

	at com.google.firebase.auth.UserRecord.checkPhotoUrl(UserRecord.java:255)
	at com.google.firebase.auth.ImportUserRecord$Builder.build(ImportUserRecord.java:262)
	at com.google.firebase.auth.ImportUserRecordTest.testInvalidPhotoUrl(ImportUserRecordTest.java:121)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.net.MalformedURLException: no protocol: not a url
	at java.net.URL.<init>(URL.java:593)
	at java.net.URL.<init>(URL.java:490)
	at java.net.URL.<init>(URL.java:439)
	at com.google.firebase.auth.UserRecord.checkPhotoUrl(UserRecord.java:253)
	... 24 more
```

File: src/main/java/com/google/firebase/auth/UserRecord.java [hunk 13]
```diff
@@ -232,29 +229,44 @@ public UpdateRequest updateRequest() {
...
+  static void checkPhoneNumber(String phoneNumber) {
     // Phone number verification is very lax here. Backend will enforce E.164 spec compliance, and
     // normalize accordingly.
     checkArgument(!Strings.isNullOrEmpty(phoneNumber), "phone number cannot be null or empty");
-    checkState(phoneNumber.startsWith("+"),
+    checkArgument(phoneNumber.startsWith("+"),
         "phone number must be a valid, E.164 compliant identifier starting with a '+' sign");
   }
 
+  static void checkPhotoUrl(String photoUrl) {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/UserRecord.java]: Done

File: src/main/java/com/google/firebase/auth/hash/BasicHash.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Map;
+
+abstract class BasicHash extends UserImportHash {
+
+  private final int rounds;
+
+  BasicHash(String name, Builder builder) {
+    super(name);
+    checkArgument(builder.rounds >= 0 && builder.rounds <= 120000,
+        "A non-empty key is required for HMAC algorithms");
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/hash/BasicHash.java]: Fixed

File: src/main/java/com/google/firebase/auth/hash/Scrypt.java (hunk 1 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/hash/Scrypt.java]: It seems there are 5 types of hash algorithms that we need to support:

1. HMAC hashes
2. MD5, SHA, PBKDF etc
3. Scrypt
4. StandardScrypt
5. Bcrypt

I refer to the 2nd category as basic hashes for the lack of a better name. I think it makes sense to not have Scrypt extend from BasicHash (the allowed integer ranges are different). But I'm certainly open to calling BasicHash something else. Any suggestions?

File: src/main/java/com/google/firebase/auth/internal/UploadAccountResponse.java (hunk 2 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/internal/UploadAccountResponse.java]: Done

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java (hunk 3 above)
Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: Done

File: src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java (hunk 4 above)
Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java]: Done

**Comment by hiranya911** [comment]
@schmidt-sebastian thanks for the thorough review. Addressed pretty much all the comments, and left one follow up question (about renaming `BasicHash`).

File: src/main/java/com/google/firebase/auth/hash/Scrypt.java (hunk 1 above)
Comment (schmidt-sebastian) [reply src/main/java/com/google/firebase/auth/hash/Scrypt.java]: What do you think of making BasicHash a "RepeatableHash" interface? You will have to verify the range in each implementing class, but it's a nicely named interface and would convey some property about the hash functions that adhere to it.

Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/hash/Scrypt.java]: Thanks for the suggestion. I renamed `BasicHash` to `RepeatableHash`, and passed range interval as constructor parameters. This enables us to reuse it in `Scrypt` as well.

File: src/main/java/com/google/firebase/auth/hash/RepeatableHash.java [hunk 15]
```diff
@@ -22,19 +22,19 @@
...
 
   private final int rounds;
 
-  BasicHash(String name, Builder builder) {
+  RepeatableHash(String name, int min, int max, Builder builder) {
     super(name);
-    checkArgument(builder.rounds >= 0 && builder.rounds <= 120000,
-        "Rounds value must be between 0 and 120000 (inclusive).");
+    checkArgument(builder.rounds >= min && builder.rounds <= max,
+        "Rounds value must be between %s and %s (inclusive).", min, max);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/hash/RepeatableHash.java]: Guava only accepts `%s`: https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Preconditions.html

File: src/main/java/com/google/firebase/auth/hash/RepeatableHash.java [hunk 16]
```diff
@@ -22,19 +22,19 @@
 import com.google.firebase.auth.UserImportHash;
 import java.util.Map;
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/hash/RepeatableHash.java]: Done

**Pull request merged**